java/arcs/core/storage/keys/JoinStorageKey.kt

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.core.storage.keys
+
+import arcs.core.storage.StorageKey
+import arcs.core.storage.StorageKeyParser
+import arcs.core.storage.referencemode.embed
+import arcs.core.storage.referencemode.unEmbed
+
+/** Protocol to be used when the StorageKey is composed of multiple StorageKeys. */
+const val COMPOSITE_PROTOCOL = "join"
+
+class JoinStorageKey(
+    private val n: Int,
+    private val storageKeys: List<StorageKey>
+) : StorageKey(COMPOSITE_PROTOCOL) {
+    override fun toKeyString(): String {
+        val builder = StringBuilder()
+        builder.append("$n/")
+        storageKeys.forEach { storageKeys ->
+            builder.append("{${storageKeys.embed()}}")
+        }
+
+        return builder.toString()
+    }
+
+    override fun childKeyWithComponent(component: String): StorageKey {
+        TODO("Not yet required for JoinStorageKey")
+    }
+
+    companion object {
+        init {
+            StorageKeyParser.addParser(COMPOSITE_PROTOCOL, ::fromString)
+        }
+
+        fun registerParser() {
+            StorageKeyParser.addParser(COMPOSITE_PROTOCOL, ::fromString)
+        }
+
+        private fun fromString(rawValue: String): JoinStorageKey {
+            val invalidFormatMessage: () -> String =
+                { "Invalid format for JoinStorageKey: $rawValue" }
+            val storageKeys = ArrayList<StorageKey>()
+
+            // We will support < 10 joins.
+            val n: Int = rawValue[0] - '0'
+            require(n in 1..9, invalidFormatMessage)
+            require(rawValue[1] == '/', invalidFormatMessage)
+            var openCount = 0
+            var openIndex = -1
+            val rawKeysValue = rawValue.substring(2)
+            rawKeysValue.forEachIndexed { i, char ->
```

Comment:
If this is very similar to what ReferenceModeStorageKey does, should we extract a utility for parsing the embedded storage keys?

Reply:
extracted a StorageKeyUtil.